Provide document of numeric types and type stability (#1938)
trixi-framework/Trixi.jl: docs/src/conventions.md

  the code.
- `wrap_array` should be used as default option. `wrap_array_native` should only
  be used when necessary, e.g., to avoid additional overhead when interfacing
  with external C libraries such as HDF5, MPI, or visualization.

## Numeric types and type stability
In Trixi.jl, `Float32` and `Float64` types are fully supported. We ensure the type stability of these numeric types throughout the development process. Below are some guidelines to apply in various scenarios.

- **Exact floating-point numbers**: Some real numbers can be represented exactly in machine (e.g., `0.25`, `0.5`, `1/2`) and we prefer to use `Float32` type of them to achieve a concise way of possible type promotion. For example,
  ```julia
  # Assume we have `0.25`, `0.5`, `1/2` in function
  0.25f0, 0.5f0, 0.5f0 # corresponding numbers
  ```

- **Non-exact floating-point numbers**: For real numbers that cannot be exactly represented in machine precision (e.g., `0.1`, `1/3`, `pi`), use the `convert` function to make them consistent with the type of the function input. For example, 
  ```julia
  # Assume we are handling `pi` in function

>   function foo(input1, input2, input3, ...)
>     RealT = eletype(input1) # good practice to always use the first input to extract numeric type
>     # ...
>     c1 = convert(RealT, pi) * c2 # sample operation
>     # ...

Reviewer: ```suggestion
  function foo(u, input2, input3, ...)
    RealT = eltype(u) # good practice to always use the first input to extract numeric type
    # ...
    c1 = convert(RealT, pi) * c2 # sample operation
    # ...
```

Maybe like this to clarify that we typically want to use the same number type as the unknowns `u`?

Author: What is the difference? The original one suggests a way to better use the first input, as otherwise people might think we should use an explicit `u` to get the number types. This could be confusing since some functions do not take `u` as input.

Reviewer: Fair enough. What about being even more explicit here since not all functions have a first argument where it would make sense to get a real type from? Concretely, we could basically leave your suggestion with the fix

```suggestion
  function foo(input1, input2, input3, ...)
    RealT = eltype(input1) # good practice to always use the first input to extract numeric type
    # ...
    c1 = convert(RealT, pi) * c2 # sample operation
    # ...
```

and list some recommendations later:

1. If the function gets a local pointwise vector of the solution variables `u` such as `flux(u, equations)`, use `u` to determine the real type `eltype(u)`.
2. If `u` is not passed as an argument but a vector of coordinates `x` such as `initial_condition(x, t, equations)`, use `eltype(x)` instead.
3. Choose an appropriate argument to determine the real type otherwise.

  end
  ```

- **Integer numbers**: Integers need special consideration. Using functions like `convert` (as mentioned above), `zero`, and `one` to change integers to a specific type or keeping them in integer format is feasible in most cases. We should balance code readability and consistency while maintaining type stability. Here are some examples to guide our developers.
  ```julia
  # The first example - `SVector`, keep code consistency within `SVector`
  SVector(0, 0, 0)
  SVector(zero(RealT), zero(RealT), zero(RealT))
  Svector(one(RealT), one(RealT), one(RealT))

  # The second example - inner functions, keep them type-stable as well

>   function foo(input1, input2, input3, ...)
>   RealT = eletype(input1) # good practice to always use the first input to extract numeric type

Reviewer: ```suggestion
  function foo(u, input1, input2, ...)
  RealT = eltype(u) # good practice to always use the first input to extract numeric type
```

  # ...
  c1 = c2 > 0.5f0 ? one(RealT) : convert(RealT, 0.1) # make type-stable
  # ...
  end 

  # The third example - some operations (e.g., `/`, `sprt`, `inv`), convert them definitely
  c1 = convert(RealT, 4) # suppose we get RealT before
  c2 = 1 / c1

>   c3 = sprt(c1)

Reviewer: ```suggestion
  c3 = sqrt(c1)
```

Author: Thanks.

  (or `wrap_array_native(u_ode, semi)`) for further processing.
- When some solution is passed together with the `mesh, equations, solver, cache, ...`,
  it is already wrapped via `wrap_array` (or `wrap_array_native`).
- Exceptions of this rule are possible, e.g., for AMR, but must be documented in
  the code.
- `wrap_array` should be used as default option. `wrap_array_native` should only
  be used when necessary, e.g., to avoid additional overhead when interfacing
  with external C libraries such as HDF5, MPI, or visualization.

## Numeric types and type stability
In Trixi.jl, `Float32` and `Float64` types are fully supported. We ensure the type stability of these numeric types throughout the development process. Below are some guidelines to apply in various scenarios.

> - **Exact floating-point numbers**: Some real numbers can be represented exactly in machine (e.g., `0.25`, `0.5`, `1/2`) and we prefer to use `Float32` type of them to achieve a concise way of possible type promotion. For example,

Reviewer: If you want to be concrete, you need to require that `m` and `n` are close enough to zero, e.g.,
```julia
julia> Float32(2)^256
Inf32

julia> Float64(2)^256
1.157920892373162e77
```
and 
```julia
julia> Float32(2)^-256
0.0f0

julia> Float64(2)^-256
8.636168555094445e-78
```

> - **Exact floating-point numbers**: Some real numbers can be represented exactly in machine (e.g., `0.25`, `0.5`, `1/2`) and we prefer to use `Float32` type of them to achieve a concise way of possible type promotion. For example,

Reviewer 2: ```suggestion
### Exact floating-point numbers

Some real numbers can be represented exactly as both `Float64` and `Float32` values (e.g., `0.25`, `0.5`, `1/2`). This is generally true for integer multiples of powers of two, i.e., numbers that can be written as ``m 2^n``, where ``m, n \in \mathbb{Z}`` and as long as ``m`` and ``n`` are not too large. We prefer to use `Float32` type for these numbers to achieve a concise way of possible type promotion. For example,
```
Good point. Maybe like this?

> - **Exact floating-point numbers**: Some real numbers can be represented exactly in machine (e.g., `0.25`, `0.5`, `1/2`) and we prefer to use `Float32` type of them to achieve a concise way of possible type promotion. For example,

Reviewer: Fine with me 👍

> - **Exact floating-point numbers**: Some real numbers can be represented exactly in machine (e.g., `0.25`, `0.5`, `1/2`) and we prefer to use `Float32` type of them to achieve a concise way of possible type promotion. For example,

Author: I suggest not being too specific, instead, a general approach is preferred here. To be concrete, `m` and `n` should fall within a specific range, i.e., they should not be too large or too small. See
https://en.wikipedia.org/wiki/Single-precision_floating-point_format 
https://en.wikipedia.org/wiki/Double-precision_floating-point_format

> - **Exact floating-point numbers**: Some real numbers can be represented exactly in machine (e.g., `0.25`, `0.5`, `1/2`) and we prefer to use `Float32` type of them to achieve a concise way of possible type promotion. For example,

Reviewer 2: I agree. Being to specific here is not helpful. However, many constants used in Trixi.jl seem to be in the range of 0.001 to 10, thus giving a workable rule of thumb when a number can be represented exactly is imho very helpful to the developers. 

Since this convo is outdated, I'll make a new suggestion and resolve this one.

>   ```julia
>   # Assume we have `0.25`, `0.5`, `1/2` in function
>   0.25f0, 0.5f0, 0.5f0 # corresponding numbers
>   ```

Reviewer 2: ```suggestion
```julia
# Assume we have `0.25`, `0.5`, `1/2` in function
0.25f0, 0.5f0, 0.5f0 # corresponding numbers
```
```

- **Non-exact floating-point numbers**: For real numbers that cannot be exactly represented in machine precision (e.g., `0.1`, `1/3`, `pi`), use the `convert` function to make them consistent with the type of the function input. For example, 
  ```julia
  # Assume we are handling `pi` in function
  function foo(..., input, ...)
    RealT = eltype(input) # see **notes** below
    # ...
    c1 = convert(RealT, pi) * c2 # sample operation
    # ...
  end

>   ```

Reviewer 2: ~~~suggestion
### Non-exact floating-point numbers

For real numbers that cannot be exactly represented in machine precision (e.g., `0.1`, `1/3`, `pi`), use the `convert` function to make them consistent with the type of the function input. For example, 
```julia
# Assume we are handling `pi` in function
function foo(..., input, ...)
  RealT = eltype(input) # see **notes** below
  # ...
  c1 = convert(RealT, pi) * c2 # sample operation
  # ...
end
```
~~~

- **Integer numbers**: Integers need special consideration. Using functions like `convert` (as mentioned above), `zero`, and `one` to change integers to a specific type or keeping them in integer format is feasible in most cases. We should balance code readability and consistency while maintaining type stability. Here are some examples to guide our developers.
  ```julia
  # The first example - `SVector`, keep code consistency within `SVector`
  SVector(0, 0, 0)
  SVector(zero(RealT), zero(RealT), zero(RealT))
  Svector(one(RealT), one(RealT), one(RealT))

  # The second example - inner functions, keep them type-stable as well
  function foo(..., input, ...)
  RealT = eltype(input) # see **notes** below
  # ...
  c1 = c2 > 0.5f0 ? one(RealT) : convert(RealT, 0.1) # make type-stable
  # ...
  end 

  # The third example - some operations (e.g., `/`, `sqrt`, `inv`), convert them definitely
  c1 = convert(RealT, 4) # suppose we get RealT before
  c2 = 1 / c1
  c3 = sqrt(c1)
  c4 = inv(c1)
  ```

>   In general, in the case of integer numbers, our developers should apply a case-by-case strategy to maintain type stability. 

Reviewer 2: ~~~suggestion
### Integer numbers

Integers need special consideration. Using functions like `convert` (as mentioned above), `zero`, and `one` to change integers to a specific type or keeping them in integer format is feasible in most cases. We aim to balance code readability and consistency while maintaining type stability. Here are some examples to guide our developers.
```julia
# The first example - `SVector`, keep code consistency within `SVector`
SVector(0, 0, 0)
SVector(zero(RealT), zero(RealT), zero(RealT))
Svector(one(RealT), one(RealT), one(RealT))

# The second example - inner functions, keep them type-stable as well
function foo(..., input, ...)
RealT = eltype(input) # see **notes** below
# ...
c1 = c2 > 0.5f0 ? one(RealT) : convert(RealT, 0.1) # make type-stable
# ...
end 

# The third example - some operations (e.g., `/`, `sqrt`, `inv`), convert them definitely
c1 = convert(RealT, 4) # suppose we get RealT before
c2 = 1 / c1
c3 = sqrt(c1)
c4 = inv(c1)
```
In general, in the case of integer numbers, our developers should apply a case-by-case strategy to maintain type stability. 
~~~

>   **Notes:** 
>   1. If the function gets a local pointwise vector of the solution variables `u` such as `flux(u, equations)`, use `u` to determine the real type `eltype(u)`.
>   2. If `u` is not passed as an argument but a vector of coordinates `x` such as `initial_condition(x, t, equations)`, use `eltype(x)` instead.
>   3. Choose an appropriate argument to determine the real type otherwise.

Reviewer 2: ```suggestion
### Notes
1. If the function gets a local pointwise vector of the solution variables `u` such as `flux(u, equations)`, use `u` to determine the real type `eltype(u)`.
2. If `u` is not passed as an argument but a vector of coordinates `x` such as `initial_condition(x, t, equations)`, use `eltype(x)` instead.
3. Choose an appropriate argument to determine the real type otherwise.
```

  (or `wrap_array_native(u_ode, semi)`) for further processing.
- When some solution is passed together with the `mesh, equations, solver, cache, ...`,
  it is already wrapped via `wrap_array` (or `wrap_array_native`).
- Exceptions of this rule are possible, e.g., for AMR, but must be documented in
  the code.
- `wrap_array` should be used as default option. `wrap_array_native` should only
  be used when necessary, e.g., to avoid additional overhead when interfacing
  with external C libraries such as HDF5, MPI, or visualization.

## Numeric types and type stability

In Trixi.jl, we use generic programming to support custom data types to store the numerical simulation data, including standard floating point types and automatic differentiation types.
Specifically, `Float32` and `Float64` types are fully supported, including the ability to run Trixi.jl on hardware that only supports `Float32` types.
We ensure the type stability of these numeric types throughout the development process.
Below are some guidelines to apply in various scenarios.

### Exact floating-point numbers

> Some real numbers can be represented exactly in machine (e.g., `0.25`, `0.5`, `1/2`) and we prefer to use `Float32` type of them to achieve a concise way of possible type promotion. For example,

Reviewer 2: ```suggestion
Some real numbers can be represented exactly as both `Float64` and `Float32` values (e.g., `0.25`, `0.5`, `1/2`). We prefer to use `Float32` type for these numbers to achieve a concise way of possible type promotion. For example,
```

```julia
# Assume we have `0.25`, `0.5`, `1/2` in function
0.25f0, 0.5f0, 0.5f0 # corresponding numbers
```

> 

Reviewer 2: ~~~suggestion
Generally, this equivalence is true for integer multiples of powers of two. That is, numbers that can be written as ``m 2^n``, where ``m, n \in \mathbb{Z}``, and where ``m`` and ``n`` are such that the result is representable as a [single precision floating point](https://en.wikipedia.org/wiki/Single-precision_floating-point_format) value. If a decimal value `v` is exactly representable in `Float32`, the expression
```julia
Float32(v) == v
```
will evaluate to `true`.

~~~

- `wrap_array` should be used as default option. `wrap_array_native` should only
  be used when necessary, e.g., to avoid additional overhead when interfacing
  with external C libraries such as HDF5, MPI, or visualization.

> # Numerical Types and Stability

Reviewer: It looks like something went wrong and the docuemntation is missing?

Author: No, I am waiting for both of you to make a final decision on how to address integer issues so that I can start the documentation. Now it is just a title. Michael seems to prefer converting each integer to the floating-point type when both options work. Please check the main PR.

Reviewer: I see. I was just confused since you requested my review on this PR.

Author: Okay. I will upload an initial version. Please wait
